# Project 2 Programming for Data Analysis Winter 2023 

**Author: Nur Bujang**

project2.ipynb
***

### Project Instruction:

> Analyse CO2 vs Temperature Anomaly from 800kyrs – present.

> Examine one other (paleo/modern) features (e.g. CH4 or polar ice-coverage)

> Examine Irish context:

    > Climate change signals: (see Maynooth study: The emergence of a climate change signal in long-term Irish meteorological observations - ScienceDirect)

> Fuse and analyse data from various data sources and format fused data set as a pandas dataframe and export to csv and json formats

> For all of the above variables, analyse the data, the trends and the relationships between them (temporal leads/lags/frequency analysis).

> Predict global temperature anomaly over next few decades (synthesise data) and compare to published climate models if atmospheric CO2 trends continue

> Comment on accelerated warming based on very latest features (e.g. temperature/polar-ice-coverage)

# Project Title: 

## **An Analysis Of Paleo-Present Climate Data**

## Abstract


## 1.0 Background

### 1.1 Climate Change

Climate is characterized by long-term patterns in temperature, humidity, winds, rainfall and extreme weather frequency. Climate change refers to significant and persistent changes in the local, regional and global climate pattern (NASA, 2023b). According to United Nations (2023) and NASA (2023a), the impacts of climate change include rising sea levels, catastrophic and frequent floods and droughts and melting polar ice. 

### 1.2 Global Warming
Global warming is a specific aspect of climate change defined by the abnormally-accelerated increase in Earth's average surface temperature due to human activity (Riebeek, 2010). This surface temperature increase has been observed since the pre-industrial period, mostly due to greenhouse gasses from fossil fuel burning (NASA, 2023b). According to United States Environmental Protection Agency (2017), examples of greenhouse gases include carbon dioxide(CO<sub>2</sub>), nitrous oxide(N<sub>2</sub>O), methane (CH<sub>4</sub>) and sulfur hexafluoride (SF<sub>6</sub>).

### 1.3 Links Between CO<sub>2</sub>, Temperature, CH<sub>4</sub> And Precipitation In Climate Change

The interrelationships between atmospheric carbon dioxide (CO2), methane (CH4), temperature, and precipitation is intricate and play a vital role in the complex system of climate change. The greenhouse gases (CO<sub>2</sub> and CH<sub>4</sub>) trap heat in the Earth's atmosphere by absorbing and re-emitting infrared radiation back toward Earth's surface (www.ces.fau.edu, n.d.). This Greenhouse Effect results in a rise in average global temperatures, impacting various components of the Earth's climate system. 

These warmer temperatures affect the atmospheric circulation and moisture patterns through intensification of the water cycle where warmer air holds more moisture (Wikipedia, 2023). This shifts in rainfall and precipitation distribution patterns leads to intense and frequent extreme weather events, such as heavy rainfall, storms, heatwaves and droughts (Understanding Global Change, n.d., Dore, 2005, www.eea.europa.eu, n.d.).

Changes in temperature and precipitation can also create an ice-albedo feedback, where the reduction Earth's albedo (reflectivity) from ice and snow melt caused by warmer temperatures leads to more absorption of solar radiation, thus further warming the planet (CIRES Education and Outreach at CU Boulder, n.d.).

#### 1.3.1 Atmospheric CO<sub>2</sub>

According to Berner (1991), atmospheric CO<sub>2</sub> dictates the world climate on geological time magnitude. Results from ice-core data show that atmospheric CO<sub>2</sub> concentrations were significantly lower between 650kyr and 750kyr compared to the present day (Lüthi et al., 2008). A few years later, Bereiter et al. (2015) followed up with a corrected 442–816kyr portion of the EPICA Dome C CO<sub>2</sub> data.

According to Hertzberg and Schreuder (2016), there has been a steady increase in the average atmospheric CO<sub>2</sub> concentration from 280ppmv to 400ppmv over the last two centuries. Data from atmospheric CO<sub>2</sub> concentration collected at  at Mauna Loa Observatory, Hawaii shows a linear trend from 315ppm in 1959 to 425ppm in 2022 (NOAA, 2023).

#### 1.3.2 Temperature

Results from high-resolution deuterium measurements of ice core data show that based on 100-year mean values over the past 800kyr, surface temperature increased ~15°C (Jouzel et al., 2007). Parrenin et al. (2013) found that atmospheric CO<sub>2</sub> and surface air temperature move synchronously for most part from 800kyr to the present time eventhough Fischer et al. (1999), Monnin et al. (2001) and Lourantou et al. (2010) reported that atmospheric CO<sub>2</sub> lags behind temperature by a few hundred years. However, Hertzberg and Schreuder (2016) noticed that temperature changes always precedes atmospheric CO<sub>2</sub> changes by several hundred to a thousand years, thus suggesting that temperature change causes CO<sub>2</sub>, and not vice versa. 

Chen et al. (2023) reported a correlation between atmospheric CO<sub>2</sub> and temperature. Earlier, Kang and Larsson (2013) reported a feedback system where temperature and CO<sub>2</sub> Granger-cause each other in both directions.

#### 1.3.3 Atmospheric CH<sub>4</sub>

Petit et al. (1999) reported that CO<sub>2</sub> and CH<sub>4</sub> concentrations follow the same trend for each glacial cycle. Monnin et al. (2001) reported that eventough the reasons behind CH<sub>4</sub> concentrations variations are different from CO<sub>2</sub>, in general, CO<sub>2</sub> and CH<sub>4</sub> increase are parallel with each other. There is also a high correlation between temperature and CH<sub>4</sub> in the past (Spahni et al., 2005). Using Granger causality analysis, temperature, CO<sub>2</sub> and CH<sub>4</sub> bidirectionally Granger-cause each other, where one's increase causes an increase in the other in a feedback system (Larsson and Persson, 2023).

#### 1.3.4 Irish Precipitation

An analysis of a homogenized Island of Ireland Precipitation (IIP) data from 1850–2010 to study long-term rainfall series in Ireland found an increasing pattern or upward trend in the amount of precipitation (rainfall or snowfall) during the winter months, while a negative or decreasing pattern was found in the amount of precipitation during the summer months (Noone et al., 2016). 

Long-term air temperature series data for Ireland between 1831-1968 was reconstructed by Mateus, Potito and Curley (2020) and between 1885–2018 (Mateus and Potito, 2021). Subsequent analysis between 1885 to 2018 found significant air temperature increase in all seasons, especially spring and autumn and significant air temperature decrease during cold days, cold nights and frost days (Mateus and Potito, 2022).

Later, Murphy et al. (2023a) found that Dublin-based stations show a maximum increase of 1.14°C in the annual mean temperature for every 1°C increase in the global mean surface temperature and increase in extremely heavy rainfall were recorded for 30% of stations. Such temperature increase would result in higher a risk of river and groundwater flooding in Ireland (Murphy et al., 2023b).

## 2.0 Methods and Implementation

### 2.1 CO<sub>2</sub> vs Temperature Anomaly From 800kyrs – Present

#### 2.1.1 CO<sub>2</sub>

Data for CO<sub>2</sub> was a composite of Chen et al. (2023) for 800kyr-2001 and Lan, Tans and Thoning (2023) for 2002-2022.

In [309]:
import pandas as pd
import numpy as np
import plotly.express as px

#### CO<sub>2</sub> pre 2001

In [310]:
# CO2 from IPCC report (Chen et al., 2023)

df_co2_pre2001 = pd.read_csv('data/co2/coipcc800kyr-2001.csv')
df_co2_pre2001.head(3)


Gasage (yr BP)   co2 (ppmv)  sigma mean CO2 (ppmv)  Unnamed: 3  Unnamed: 4  \
0           -51.03      368.02                   0.06         NaN         NaN   
1           -48.00      361.78                   0.37         NaN         NaN   
2           -46.28      359.65                   0.10         NaN         NaN   

      year  
0  2001.03  
1  1998.00  
2  1996.28

In [311]:
df_co2_pre2001 = df_co2_pre2001.iloc[:, :2].copy()
df_co2_pre2001.head(3)

Gasage (yr BP)   co2 (ppmv)
0           -51.03      368.02
1           -48.00      361.78
2           -46.28      359.65

In [312]:
df_co2_pre2001['year'] = (np.negative(df_co2_pre2001['Gasage (yr BP) '])
                          + 1950)
df_co2_pre2001.head(3)

Gasage (yr BP)   co2 (ppmv)     year
0           -51.03      368.02  2001.03
1           -48.00      361.78  1998.00
2           -46.28      359.65  1996.28

In [313]:
df_co2_pre2001.rename(columns={'co2 (ppmv)': 'co2 (ppmv) ipcc'}, 
                    inplace=True)
df_co2_pre2001.head(3)

Gasage (yr BP)   co2 (ppmv) ipcc     year
0           -51.03           368.02  2001.03
1           -48.00           361.78  1998.00
2           -46.28           359.65  1996.28

In [314]:
df_co2_pre2001 = df_co2_pre2001[['year','co2 (ppmv) ipcc']].copy()
df_co2_pre2001.head(3)

year  co2 (ppmv) ipcc
0  2001.03           368.02
1  1998.00           361.78
2  1996.28           359.65

In [315]:
# CO2 from Lüthi et al. (2008)

df_co2_luthi = pd.read_csv('data/co2/coluthi.csv')
df_co2_luthi.head(3)

EDC3_gas_a (yr)  co2 (ppmv)
0              137       280.4
1              268       274.9
2              279       277.9

In [316]:
df_co2_luthi.rename(columns={'co2 (ppmv)': 'co2 (ppmv) luthi'}, 
                    inplace=True)
df_co2_luthi.head(3)

EDC3_gas_a (yr)  co2 (ppmv) luthi
0              137             280.4
1              268             274.9
2              279             277.9

In [317]:
df_co2_luthi['year'] = (np.negative(df_co2_luthi['EDC3_gas_a (yr)'])
                          + 1950)
df_co2_luthi.head(3)

EDC3_gas_a (yr)  co2 (ppmv) luthi  year
0              137             280.4  1813
1              268             274.9  1682
2              279             277.9  1671

In [318]:
df_co2_luthi = df_co2_luthi[['year','co2 (ppmv) luthi']].copy()
df_co2_luthi.head(3)

year  co2 (ppmv) luthi
0  1813             280.4
1  1682             274.9
2  1671             277.9

In [319]:
df_co2_ipcc_vs_luthi = pd.concat([df_co2_pre2001, 
                        df_co2_luthi
])
df_co2_ipcc_vs_luthi

year  co2 (ppmv) ipcc  co2 (ppmv) luthi
0       2001.03           368.02               NaN
1       1998.00           361.78               NaN
2       1996.28           359.65               NaN
3       1994.41           357.11               NaN
4       1993.08           353.95               NaN
...         ...              ...               ...
1091 -792658.00              NaN             199.4
1092 -793252.00              NaN             195.2
1093 -794517.00              NaN             189.3
1094 -795149.00              NaN             188.4
1095 -796562.00              NaN             191.0

[2997 rows x 3 columns]

In [320]:
px.scatter(df_co2_ipcc_vs_luthi.sort_values('year', ascending=True), 
        x='year', 
        y=['co2 (ppmv) ipcc', 'co2 (ppmv) luthi'], 
        width=1200,
        title="Comparison of CO2 Dataset from IPCC Report and Luthi et al. 2008"
        )


The CO<sub>2</sub> dataset from Luthi et al. 2008 is not a subset of the IPCC Report. Values between 17.80kyr and 111.86kyr are different from one another. Similarities in CO<sub>2</sub> values between the two dataset suggests a gas age revision in IPCC Report, as seen from 154.5-218.2kyr, 332.9-389.9kyr and 689.6-800kyr.

#### CO<sub>2</sub> post 2001

In [321]:
# Standardise CO2 columns

df_co2_pre2001.rename(columns={'co2 (ppmv) ipcc': 'co2 (ppmv)'}, 
                    inplace=True)

df_co2_luthi.rename(columns={'co2 (ppmv) luthi': 'co2 (ppmv)'}, 
                    inplace=True)

In [322]:
df_co2_post2001 = pd.read_csv('data/co2/conoaa1979-2022.csv', skiprows=37)
df_co2_post2001.head(3)

year    mean   unc
0  1979  336.85  0.11
1  1980  338.91  0.07
2  1981  340.11  0.09

In [323]:
df_co2_post2001 = df_co2_post2001.iloc[:, :2].copy()
df_co2_post2001.rename(columns={'mean': 'co2 (ppmv)'}, inplace=True)
df_co2_post2001.tail(3)

year  co2 (ppmv)
41  2020      412.44
42  2021      414.70
43  2022      417.07

#### CO<sub>2</sub> complete

In [324]:
df_co2_all = pd.concat([df_co2_pre2001, 
                        df_co2_luthi, 
                        df_co2_post2001
])
df_co2_all

# *********************** concat not the best way to merge


year  co2 (ppmv)
0   2001.03      368.02
1   1998.00      361.78
2   1996.28      359.65
3   1994.41      357.11
4   1993.08      353.95
..      ...         ...
39  2018.00      407.61
40  2019.00      410.07
41  2020.00      412.44
42  2021.00      414.70
43  2022.00      417.07

[3041 rows x 2 columns]

In [325]:
df_co2_all.sort_values('year', ascending=True)
df_co2_all

year  co2 (ppmv)
0   2001.03      368.02
1   1998.00      361.78
2   1996.28      359.65
3   1994.41      357.11
4   1993.08      353.95
..      ...         ...
39  2018.00      407.61
40  2019.00      410.07
41  2020.00      412.44
42  2021.00      414.70
43  2022.00      417.07

[3041 rows x 2 columns]

In [326]:
df_co2_all.to_csv('allco2.csv', index=False)

#### 2.1.2 Temperature

Data for temperature was a composite of National Centers for Environmental Information (NOAA NCEI, n.d.), Jouzel and Masson‐Delmotte (2007) versions of Jouzel et al. (2007) for 800kyr-2000 and NOAA NCEI (2023b) for 2001-2022, where the 20th century global average temperature is 13.9°C (57.0°F) NOAA NCEI (2023a).

In the tempnoaajoudelt2000-801662 dataset (Jouzel and Masson‐Delmotte, 2007), $\delta$ T is the temperature difference from the average of the last 1000 years in °C, with corrections (Bintanja, van de Wal and Oerlemans, 2005, Parrenin et al., 2007). The dataset tempjouzelaicc2000-800kyr from AICC12 has a different ice core chronology and timescale.

find baseline/is there a constant from jouzel to get the noaa anomaly
subtract one graph from another



#### Temperature pre 2005

In [327]:
# Temperature from Jouzel-AICC

df_temp_joai = pd.read_csv('data/at/tempjouzelaicc2005-800kyr.csv')
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal   temp  acc-EDC3beta
0    1  0.00 -50.00000 -55.000000        NaN  218.3      3.105321
1    2  0.55 -43.55977 -50.513333        NaN  218.3      3.104805
2    3  1.10 -37.44019 -46.026667        NaN  218.3      3.104404

In [328]:
df_temp_joai.rename(columns={'temp': 'temp k'}, inplace=True)
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal  temp k  acc-EDC3beta
0    1  0.00 -50.00000 -55.000000        NaN   218.3      3.105321
1    2  0.55 -43.55977 -50.513333        NaN   218.3      3.104805
2    3  1.10 -37.44019 -46.026667        NaN   218.3      3.104404

In [329]:
df_temp_joai['temp c'] = df_temp_joai['temp k'] - 273.15
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal  temp k  acc-EDC3beta  temp c
0    1  0.00 -50.00000 -55.000000        NaN   218.3      3.105321  -54.85
1    2  0.55 -43.55977 -50.513333        NaN   218.3      3.104805  -54.85
2    3  1.10 -37.44019 -46.026667        NaN   218.3      3.104404  -54.85

In [330]:
df_temp_joai['edc year'] = (np.negative(df_temp_joai['EDC3béta'])
                          + 1950)
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal  temp k  acc-EDC3beta  temp c  \
0    1  0.00 -50.00000 -55.000000        NaN   218.3      3.105321  -54.85   
1    2  0.55 -43.55977 -50.513333        NaN   218.3      3.104805  -54.85   
2    3  1.10 -37.44019 -46.026667        NaN   218.3      3.104404  -54.85   

     edc year  
0  2000.00000  
1  1993.55977  
2  1987.44019

In [331]:
df_temp_joai['aicc year'] = (np.negative(df_temp_joai['AICC2012'])
                          + 1950)
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal  temp k  acc-EDC3beta  temp c  \
0    1  0.00 -50.00000 -55.000000        NaN   218.3      3.105321  -54.85   
1    2  0.55 -43.55977 -50.513333        NaN   218.3      3.104805  -54.85   
2    3  1.10 -37.44019 -46.026667        NaN   218.3      3.104404  -54.85   

     edc year    aicc year  
0  2000.00000  2005.000000  
1  1993.55977  2000.513333  
2  1987.44019  1996.026667

In [332]:
px.scatter(df_temp_joai, x=['edc year', 'aicc year'], y='temp c', width=1200, 
           title="Comparison of Temperature Data from Jouzel et al. (2007)-AICC and NOAA NCEI (2023b)"
           )

Time (Gas age) adjustment in AICC2012 shifted the temperature values from 400kyr to 800kyr. The temperature and time from 400kyr to present remain the same.

Henceforth, gas age from AICC will be used as time unit.

In [333]:
df_temp_joai.rename(columns={'aicc year': 'year'}, inplace=True)
df_temp_joai.head(3)

bag  ztop  EDC3béta   AICC2012  deutfinal  temp k  acc-EDC3beta  temp c  \
0    1  0.00 -50.00000 -55.000000        NaN   218.3      3.105321  -54.85   
1    2  0.55 -43.55977 -50.513333        NaN   218.3      3.104805  -54.85   
2    3  1.10 -37.44019 -46.026667        NaN   218.3      3.104404  -54.85   

     edc year         year  
0  2000.00000  2005.000000  
1  1993.55977  2000.513333  
2  1987.44019  1996.026667

In [334]:
df_temp_joai.sort_values(['year'], inplace=True)

In [335]:
# Temperature from NOAA (NOAA NCEI, n.d.)

df_temp_noaanom = pd.read_csv('data/at/tempnoaajoudelt2000-801662.txt', skiprows = 91)
df_temp_noaanom.head(3)

Bag         ztop          Age         Deuterium    anom c
0                    1            0        -50.00000        
1                    2         0.55        -43.54769        
2                    3          1.1        -37.41829

In [336]:
columns = df_temp_noaanom.columns.str.split()[0][:-1]
columns


['Bag', 'ztop', 'Age', 'Deuterium', 'anom']

In [337]:
anom_split = df_temp_noaanom[' Bag         ztop          Age         Deuterium    anom c'].str.split()
anom_split.tail(3)

5797    [5798, 3188.35, 799501, -441.1, -8.88]
5798    [5799, 3188.9, 800589, -441.42, -8.92]
5799    [5800, 3189.45, 801662, -440.9, -8.82]
Name:  Bag         ztop          Age         Deuterium    anom c, dtype: object

In [338]:
for index, col in enumerate(columns):
    
    if col == 'anom':
        col = 'anom c'
    
    df_temp_noaanom[col] = anom_split.str[index]

In [339]:
df_temp_noaanom = df_temp_noaanom.iloc[:, 1:].copy()
df_temp_noaanom = df_temp_noaanom.astype(float)
df_temp_noaanom

Bag     ztop           Age  Deuterium  anom c
0        1.0     0.00     -50.00000        NaN     NaN
1        2.0     0.55     -43.54769        NaN     NaN
2        3.0     1.10     -37.41829        NaN     NaN
3        4.0     1.65     -31.61153        NaN     NaN
4        5.0     2.20     -24.51395        NaN     NaN
...      ...      ...           ...        ...     ...
5795  5796.0  3187.25  797408.00000    -440.20   -8.73
5796  5797.0  3187.80  798443.00000    -439.00   -8.54
5797  5798.0  3188.35  799501.00000    -441.10   -8.88
5798  5799.0  3188.90  800589.00000    -441.42   -8.92
5799  5800.0  3189.45  801662.00000    -440.90   -8.82

[5800 rows x 5 columns]

In [340]:
df_temp_noaanom['year'] = (np.negative(df_temp_noaanom['Age']) + 1950)
df_temp_noaanom.head(3)

Bag  ztop       Age  Deuterium  anom c        year
0  1.0  0.00 -50.00000        NaN     NaN  2000.00000
1  2.0  0.55 -43.54769        NaN     NaN  1993.54769
2  3.0  1.10 -37.41829        NaN     NaN  1987.41829

#### Temperature post 2005

In [341]:
df_temp_post2005 = pd.read_csv('data/at/tempglobalanomnoaa1850-2022.csv', skiprows = 4)
df_temp_post2005.head(3)

Year  Anomaly
0  1850    -0.37
1  1851    -0.12
2  1852    -0.15

In [342]:
df_temp_post2005.rename(columns={'Year': 'year', 
                             'Anomaly': 'anom c'}, 
                             inplace=True)
df_temp_post2005.head(3)

year  anom c
0  1850   -0.37
1  1851   -0.12
2  1852   -0.15

In [343]:
# Because the temperature average for 1901-2000 = 13.9°C:

df_temp_post2005['temp c'] = df_temp_post2005['anom c'] + 13.9
df_temp_post2005.head(3)

year  anom c  temp c
0  1850   -0.37   13.53
1  1851   -0.12   13.78
2  1852   -0.15   13.75

#### Temperature complete

In [344]:
df_temp_all = [df_temp_joai[['year', 'temp c']],
               df_temp_noaanom[['year', 'anom c']], 
               df_temp_post2005.rename(columns={'anom c': 'modern anom c', 'temp c': 'modern temp c'} )
               ]

In [345]:
df_temp_all[0]['year'] = df_temp_all[0]['year'].round(0).astype(int)

/tmp/ipykernel_5776/405913841.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [346]:
df_temp_all[1]['year'] = df_temp_all[1]['year'].round(0).astype(int)

/tmp/ipykernel_5776/1805256230.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [347]:
df_temp_all_0_1 = df_temp_all[0].merge(df_temp_all[1], 
                    left_on='year', right_on='year', 
                    how='outer')

In [348]:
df_temp_all_0_2 = df_temp_all_0_1.merge(df_temp_all[2], left_on='year', right_on='year', how='outer')
df_temp_all_0_2.sort_values('year', inplace=True)
df_temp_all_0_2.reset_index(drop=True, inplace=True)
df_temp_all_0_2.head(3)

year  temp c  anom c  modern anom c  modern temp c
0 -819888  -61.75     NaN            NaN            NaN
1 -818819  -61.75     NaN            NaN            NaN
2 -817750  -61.85     NaN            NaN            NaN

In [349]:
df_temp_all_0_2['year reset'] = df_temp_all_0_2['year'] + np.negative(df_temp_all_0_2['year'].min())
df_temp_all_0_2.head(3)

year  temp c  anom c  modern anom c  modern temp c  year reset
0 -819888  -61.75     NaN            NaN            NaN           0
1 -818819  -61.75     NaN            NaN            NaN        1069
2 -817750  -61.85     NaN            NaN            NaN        2138

In [350]:
df_temp_all_0_2.set_index('year reset', inplace=True)
df_temp_all_0_2

year  temp c  anom c  modern anom c  modern temp c
year reset                                                      
0          -819888  -61.75     NaN            NaN            NaN
1069       -818819  -61.75     NaN            NaN            NaN
2138       -817750  -61.85     NaN            NaN            NaN
3207       -816681  -61.95     NaN            NaN            NaN
4276       -815612  -62.15     NaN            NaN            NaN
...            ...     ...     ...            ...            ...
821906        2018     NaN     NaN           0.83          14.73
821907        2019     NaN     NaN           0.91          14.81
821908        2020     NaN     NaN           1.13          15.03
821909        2021     NaN     NaN           0.83          14.73
821910        2022     NaN     NaN           0.93          14.83

[11657 rows x 5 columns]

In [351]:
df_temp_all_0_2.to_excel('df_temp_all_0-2_last.xlsx')

Because pandas datetime limit is 584 years using a 64-bit integer (calmcode.io, n.d.), 

In [352]:
df = pd.read_excel('df_temp_all_0-2_last.xlsx')
df.head(3)

year reset    year  temp c  anom c  modern anom c  modern temp c
0           0 -819888  -61.75     NaN            NaN            NaN
1        1069 -818819  -61.75     NaN            NaN            NaN
2        2138 -817750  -61.85     NaN            NaN            NaN

In [353]:
df['year reset dt'] = df['year reset'].apply(lambda x: pd.Period(year=x, freq='Y'))
df.head(3)

year reset    year  temp c  anom c  modern anom c  modern temp c  \
0           0 -819888  -61.75     NaN            NaN            NaN   
1        1069 -818819  -61.75     NaN            NaN            NaN   
2        2138 -817750  -61.85     NaN            NaN            NaN   

  year reset dt  
0             0  
1          1069  
2          2138

In [354]:
df.set_index('year reset dt', inplace=True, drop=True)
df

year reset    year  temp c  anom c  modern anom c  \
year reset dt                                                      
0                       0 -819888  -61.75     NaN            NaN   
1069                 1069 -818819  -61.75     NaN            NaN   
2138                 2138 -817750  -61.85     NaN            NaN   
3207                 3207 -816681  -61.95     NaN            NaN   
4276                 4276 -815612  -62.15     NaN            NaN   
...                   ...     ...     ...     ...            ...   
821906             821906    2018     NaN     NaN           0.83   
821907             821907    2019     NaN     NaN           0.91   
821908             821908    2020     NaN     NaN           1.13   
821909             821909    2021     NaN     NaN           0.83   
821910             821910    2022     NaN     NaN           0.93   

               modern temp c  
year reset dt                 
0                        NaN  
1069                     NaN  
2138                     NaN  
3207                     NaN  
4276                     NaN  
...                      ...  
821906                 14.73  
821907                 14.81  
821908                 15.03  
821909                 14.73  
821910                 14.83  

[11657 rows x 6 columns]

In [355]:
resampled = df.resample('1000Y')['temp c'].mean().to_frame(name='temp c mean')

In [356]:
resampled_float = resampled.reset_index().astype(str).astype(float)

In [357]:
px.scatter(resampled_float, x='year reset dt', y='temp c mean'
           )

### 2.2 CH<sub>4</sub> Anomaly From 800kyrs – Present

Data for CH<sub>4</sub> was a composite of Loulergue et al. (2008) for 800kyr-1937, Rubino et al. (2019) for 1938-1983 and Lan, Thoning and Dlugokencky (2023) for 1984-2022.



#### CH<sub>4</sub> pre 1937

In [358]:
df_ch4_pre1937 = pd.read_csv('data/ch4/cheloulergue800kyr-1937.txt', skiprows = 153)
df_ch4_pre1937.head(3)

Depth      Gas Age   CH4 mean   1s    Lab.
0  99.34           13     907      10      b
1  102.45         126     784      10      g
2  102.58         130     762      10      b

In [359]:
columns = ['Depth', 'Gas Age', 'CH4 mean', '1s', 'Lab.']

In [360]:
df_ch4_pre1937_split = df_ch4_pre1937['Depth      Gas Age   CH4 mean   1s    Lab.'].str.split()
df_ch4_pre1937_split.head(3)

0      [99.34, 13, 907, 10, b]
1    [102.45, 126, 784, 10, g]
2    [102.58, 130, 762, 10, b]
Name: Depth      Gas Age   CH4 mean   1s    Lab., dtype: object

In [361]:
for index, col in enumerate(columns):
    
    if col == 'CH4 mean':
        col = 'ch4 (ppbv)'
    
    df_ch4_pre1937[col] = df_ch4_pre1937_split.str[index]
    
df_ch4_pre1937 = df_ch4_pre1937[['Gas Age', 'ch4 (ppbv)']].astype(float)
df_ch4_pre1937.head(3)

Gas Age  ch4 (ppbv)
0     13.0       907.0
1    126.0       784.0
2    130.0       762.0

#### CH<sub>4</sub> 1938 - 1996

In [362]:
df_ch4_1938_1996 = pd.read_csv('data/ch4/chrubino630-1996.txt', skiprows = 110)
df_ch4_1938_1996.head(3)

SampleID\tage_ice\tage_CH4\tCH4ppb\tCH4err
0          DSSW20K 15.8\tfirn\t1996\t1707\t1
1            DSSW20K 29\tfirn\t1994\t1696\t2
2              DE08-2 0\tfirn\t1993\t1674\t1

The dataset chrubino630-1996.txt is tab-delimited.

In [363]:
df_ch4_1938_1996 = pd.read_csv('data/ch4/chrubino630-1996.txt', skiprows = 110, delimiter='\t')
df_ch4_1938_1996.head(3)

SampleID age_ice  age_CH4  CH4ppb  CH4err
0  DSSW20K 15.8    firn     1996    1707       1
1    DSSW20K 29    firn     1994    1696       2
2      DE08-2 0    firn     1993    1674       1

In [364]:
df_ch4_1938_1996 = df_ch4_1938_1996[['age_CH4', 'CH4ppb']].copy()
df_ch4_1938_1996.head(3)

age_CH4  CH4ppb
0     1996    1707
1     1994    1696
2     1993    1674

In [365]:
df_ch4_1938_1996.rename(columns = {'age_CH4':'year', 'CH4ppb':'ch4 ppbv'}, inplace=True)
df_ch4_1938_1996.head(3)

year  ch4 ppbv
0  1996      1707
1  1994      1696
2  1993      1674

#### CH<sub>4</sub> post 1997

In [366]:
df_ch4_1997_2022 = pd.read_csv('data/ch4/chnoaa1984-2022.csv', skiprows = 43)
df_ch4_1997_2022.head(3)

year     mean   unc
0  1984  1644.85  0.67
1  1985  1657.29  0.59
2  1986  1670.09  0.74

In [367]:
df_ch4_1997_2022.rename(columns={'mean': 'ch4 (ppbv)'}, inplace=True)
df_ch4_1997_2022.head(3)

year  ch4 (ppbv)   unc
0  1984     1644.85  0.67
1  1985     1657.29  0.59
2  1986     1670.09  0.74

In [368]:
df_ch4_1997_2022 = df_ch4_1997_2022[['year','ch4 (ppbv)']].copy()
df_ch4_1997_2022.head(3)

year  ch4 (ppbv)
0  1984     1644.85
1  1985     1657.29
2  1986     1670.09

#### CH<sub>4</sub> complete

### 2.3 Irish Context: Climate Change Signal

Temperature data was a composite of Mateus (2021), Mateus and Potito (2021), Mateus, Potito and Curley (2020a), Mateus, Potito and Curley (2020b) for 1831-1852 at Phoenix Park, Dublin, 1853-1854 at Botanic Gardens Glasnevin, Dublin, 1855-1940 at Phoenix Park, Dublin, and Met Eireann (2019) for 1941-2022 at Dublin Airport.

Precipitation data was a composite of Murphy et al. (2018a,b) for 1711–2016 and Met Eireann (2019) for 2017-2022 at Dublin Airport.

#### 2.3.1 Ireland Temperature

#### Temperature 1831-1852

In [369]:
df_irtemp_1831_1852 = pd.read_csv('data/ire/irtempphoepk1831-1958.csv', 
                                  encoding_errors='ignore')
df_irtemp_1831_1852.head(3)

Year  Month  Day  Max (F)  Min (F)  Max (C)  Min (C)
0  1831      1    1      NaN      NaN      NaN      NaN
1  1831      1    2      NaN      NaN      NaN      NaN
2  1831      1    3      NaN      NaN      NaN      NaN

In [370]:
df_irtemp_1831_1852.rename(columns={'Year': 'year',
                            'Month': 'month',
                            'Day': 'day',
                            'Max (C)': 'max c',
                            'Min (C)': 'min c',},
                            inplace=True)
df_irtemp_1831_1852

year  month  day  Max (F)  Min (F)  max c  min c
0      1831      1    1      NaN      NaN    NaN    NaN
1      1831      1    2      NaN      NaN    NaN    NaN
2      1831      1    3      NaN      NaN    NaN    NaN
3      1831      1    4      NaN      NaN    NaN    NaN
4      1831      1    5      NaN      NaN    NaN    NaN
...     ...    ...  ...      ...      ...    ...    ...
46746  1958     12   27     53.6     44.8   12.0    7.1
46747  1958     12   28     54.0     41.3   12.2    5.2
46748  1958     12   29     47.4     37.7    8.6    3.2
46749  1958     12   30     48.3     39.1    9.1    3.9
46750  1958     12   31     43.7     33.8    6.5    1.0

[46751 rows x 7 columns]

In [371]:
df_irtemp_1831_1852['temp c'] = df_irtemp_1831_1852['max c'] + df_irtemp_1831_1852['min c'] / 2
df_irtemp_1831_1852

year  month  day  Max (F)  Min (F)  max c  min c  temp c
0      1831      1    1      NaN      NaN    NaN    NaN     NaN
1      1831      1    2      NaN      NaN    NaN    NaN     NaN
2      1831      1    3      NaN      NaN    NaN    NaN     NaN
3      1831      1    4      NaN      NaN    NaN    NaN     NaN
4      1831      1    5      NaN      NaN    NaN    NaN     NaN
...     ...    ...  ...      ...      ...    ...    ...     ...
46746  1958     12   27     53.6     44.8   12.0    7.1   15.55
46747  1958     12   28     54.0     41.3   12.2    5.2   14.80
46748  1958     12   29     47.4     37.7    8.6    3.2   10.20
46749  1958     12   30     48.3     39.1    9.1    3.9   11.05
46750  1958     12   31     43.7     33.8    6.5    1.0    7.00

[46751 rows x 8 columns]

In [372]:
df_irtemp_1831_1852['dmy'] = (df_irtemp_1831_1852['day'].astype(str) + '-' + 
                               df_irtemp_1831_1852['month'].astype(str) + '-' + 
                               df_irtemp_1831_1852['year'].astype(str))
df_irtemp_1831_1852

year  month  day  Max (F)  Min (F)  max c  min c  temp c         dmy
0      1831      1    1      NaN      NaN    NaN    NaN     NaN    1-1-1831
1      1831      1    2      NaN      NaN    NaN    NaN     NaN    2-1-1831
2      1831      1    3      NaN      NaN    NaN    NaN     NaN    3-1-1831
3      1831      1    4      NaN      NaN    NaN    NaN     NaN    4-1-1831
4      1831      1    5      NaN      NaN    NaN    NaN     NaN    5-1-1831
...     ...    ...  ...      ...      ...    ...    ...     ...         ...
46746  1958     12   27     53.6     44.8   12.0    7.1   15.55  27-12-1958
46747  1958     12   28     54.0     41.3   12.2    5.2   14.80  28-12-1958
46748  1958     12   29     47.4     37.7    8.6    3.2   10.20  29-12-1958
46749  1958     12   30     48.3     39.1    9.1    3.9   11.05  30-12-1958
46750  1958     12   31     43.7     33.8    6.5    1.0    7.00  31-12-1958

[46751 rows x 9 columns]

In [373]:
df_irtemp_1831_1852['dmy'] = pd.to_datetime(df_irtemp_1831_1852['dmy'], format='mixed')

df_irtemp_1831_1852['dmy float'] =  np.where(df_irtemp_1831_1852['dmy'].dt.is_leap_year == True,
                                              
                                        # normal year, 365 days
                                        df_irtemp_1831_1852['dmy'].dt.year + df_irtemp_1831_1852['dmy'].dt.day_of_year / 365, 

                                        # 366 for a leap year
                                        df_irtemp_1831_1852['dmy'].dt.year + df_irtemp_1831_1852['dmy'].dt.day_of_year / 366 
                                        )
df_irtemp_1831_1852.head(3) 

year  month  day  Max (F)  Min (F)  max c  min c  temp c        dmy  \
0  1831      1    1      NaN      NaN    NaN    NaN     NaN 1831-01-01   
1  1831      1    2      NaN      NaN    NaN    NaN     NaN 1831-02-01   
2  1831      1    3      NaN      NaN    NaN    NaN     NaN 1831-03-01   

     dmy float  
0  1831.002732  
1  1831.087432  
2  1831.163934

#### Temperature 1853-1854

In [374]:
df_irtemp_1853_1854 = pd.read_csv('data/ire/irtempbotgdn1834-1958.csv', 
                                  encoding_errors='ignore')
df_irtemp_1853_1854.head(3)

Year  Month  Day  Max (F)  Min (F)  Max (C)  Min (C)
0  1834      1    1     41.0     39.0      5.0      3.9
1  1834      1    2     38.0     40.0      3.3      4.4
2  1834      1    3     52.0     50.0     11.1     10.0

In [375]:
df_irtemp_1853_1854.rename(columns={'Year': 'year',
                            'Month': 'month',
                            'Day': 'day',
                            'Max (C)': 'max c',
                            'Min (C)': 'min c',},
                            inplace=True)
df_irtemp_1853_1854.head(3)

year  month  day  Max (F)  Min (F)  max c  min c
0  1834      1    1     41.0     39.0    5.0    3.9
1  1834      1    2     38.0     40.0    3.3    4.4
2  1834      1    3     52.0     50.0   11.1   10.0

In [376]:
df_irtemp_1853_1854['temp c'] = df_irtemp_1853_1854['max c'] + df_irtemp_1853_1854['min c'] / 2
df_irtemp_1853_1854.head(3)

year  month  day  Max (F)  Min (F)  max c  min c  temp c
0  1834      1    1     41.0     39.0    5.0    3.9    6.95
1  1834      1    2     38.0     40.0    3.3    4.4    5.50
2  1834      1    3     52.0     50.0   11.1   10.0   16.10

In [377]:
df_irtemp_1853_1854['dmy'] = (df_irtemp_1853_1854['day'].astype(str) + '-' + 
                               df_irtemp_1853_1854['month'].astype(str) + '-' + 
                               df_irtemp_1853_1854['year'].astype(str))
df_irtemp_1853_1854

year  month  day  Max (F)  Min (F)  max c  min c  temp c         dmy
0      1834      1    1     41.0     39.0    5.0    3.9    6.95    1-1-1834
1      1834      1    2     38.0     40.0    3.3    4.4    5.50    2-1-1834
2      1834      1    3     52.0     50.0   11.1   10.0   16.10    3-1-1834
3      1834      1    4     50.0     48.0   10.0    8.9   14.45    4-1-1834
4      1834      1    5     50.0     49.0   10.0    9.4   14.70    5-1-1834
...     ...    ...  ...      ...      ...    ...    ...     ...         ...
45650  1958     12   27     53.6     44.0   12.0    6.7   15.35  27-12-1958
45651  1958     12   28     54.8     41.8   12.7    5.4   15.40  28-12-1958
45652  1958     12   29     48.0     38.4    8.9    3.6   10.70  29-12-1958
45653  1958     12   30     47.6     40.4    8.7    4.7   11.05  30-12-1958
45654  1958     12   31     45.0     34.4    7.2    1.3    7.85  31-12-1958

[45655 rows x 9 columns]

In [378]:
df_irtemp_1853_1854['dmy'] = pd.to_datetime(df_irtemp_1853_1854['dmy'], format='mixed')

df_irtemp_1853_1854['dmy float'] =  np.where(df_irtemp_1853_1854['dmy'].dt.is_leap_year == True,
                                              
                                        # normal year, 365 days
                                        df_irtemp_1853_1854['dmy'].dt.year + df_irtemp_1853_1854['dmy'].dt.day_of_year / 365, 

                                        # 366 for a leap year
                                        df_irtemp_1853_1854['dmy'].dt.year + df_irtemp_1853_1854['dmy'].dt.day_of_year / 366 
                                        )
df_irtemp_1853_1854.head(3) 

year  month  day  Max (F)  Min (F)  max c  min c  temp c        dmy  \
0  1834      1    1     41.0     39.0    5.0    3.9    6.95 1834-01-01   
1  1834      1    2     38.0     40.0    3.3    4.4    5.50 1834-02-01   
2  1834      1    3     52.0     50.0   11.1   10.0   16.10 1834-03-01   

     dmy float  
0  1834.002732  
1  1834.087432  
2  1834.163934

#### Temperature 1855-1940

In [379]:
df_irtemp_1855_1940 = pd.read_csv('data/ire/irtempphoepk1831-1958.csv', 
                                  encoding_errors='ignore')
df_irtemp_1855_1940.head(3)

Year  Month  Day  Max (F)  Min (F)  Max (C)  Min (C)
0  1831      1    1      NaN      NaN      NaN      NaN
1  1831      1    2      NaN      NaN      NaN      NaN
2  1831      1    3      NaN      NaN      NaN      NaN

In [380]:
df_irtemp_1855_1940.rename(columns={'Year': 'year',
                            'Month': 'month',
                            'Day': 'day',
                            'Max (C)': 'max c',
                            'Min (C)': 'min c',},
                            inplace=True)
df_irtemp_1855_1940.head(3)

year  month  day  Max (F)  Min (F)  max c  min c
0  1831      1    1      NaN      NaN    NaN    NaN
1  1831      1    2      NaN      NaN    NaN    NaN
2  1831      1    3      NaN      NaN    NaN    NaN

In [381]:
df_irtemp_1855_1940['temp c'] = df_irtemp_1855_1940['max c'] + df_irtemp_1855_1940['min c'] / 2
df_irtemp_1855_1940.head(3)

year  month  day  Max (F)  Min (F)  max c  min c  temp c
0  1831      1    1      NaN      NaN    NaN    NaN     NaN
1  1831      1    2      NaN      NaN    NaN    NaN     NaN
2  1831      1    3      NaN      NaN    NaN    NaN     NaN

In [382]:
df_irtemp_1855_1940['dmy'] = (df_irtemp_1855_1940['day'].astype(str) + '-' + 
                               df_irtemp_1855_1940['month'].astype(str) + '-' + 
                               df_irtemp_1855_1940['year'].astype(str))
df_irtemp_1855_1940

year  month  day  Max (F)  Min (F)  max c  min c  temp c         dmy
0      1831      1    1      NaN      NaN    NaN    NaN     NaN    1-1-1831
1      1831      1    2      NaN      NaN    NaN    NaN     NaN    2-1-1831
2      1831      1    3      NaN      NaN    NaN    NaN     NaN    3-1-1831
3      1831      1    4      NaN      NaN    NaN    NaN     NaN    4-1-1831
4      1831      1    5      NaN      NaN    NaN    NaN     NaN    5-1-1831
...     ...    ...  ...      ...      ...    ...    ...     ...         ...
46746  1958     12   27     53.6     44.8   12.0    7.1   15.55  27-12-1958
46747  1958     12   28     54.0     41.3   12.2    5.2   14.80  28-12-1958
46748  1958     12   29     47.4     37.7    8.6    3.2   10.20  29-12-1958
46749  1958     12   30     48.3     39.1    9.1    3.9   11.05  30-12-1958
46750  1958     12   31     43.7     33.8    6.5    1.0    7.00  31-12-1958

[46751 rows x 9 columns]

In [383]:
df_irtemp_1855_1940['dmy'] = pd.to_datetime(df_irtemp_1855_1940['dmy'], format='mixed')

df_irtemp_1855_1940['dmy float'] =  np.where(df_irtemp_1855_1940['dmy'].dt.is_leap_year == True,
                                              
                                        # normal year, 365 days
                                        df_irtemp_1855_1940['dmy'].dt.year + df_irtemp_1853_1854['dmy'].dt.day_of_year / 365, 

                                        # 366 for a leap year
                                        df_irtemp_1855_1940['dmy'].dt.year + df_irtemp_1853_1854['dmy'].dt.day_of_year / 366 
                                        )
df_irtemp_1855_1940.head(3) 

year  month  day  Max (F)  Min (F)  max c  min c  temp c        dmy  \
0  1831      1    1      NaN      NaN    NaN    NaN     NaN 1831-01-01   
1  1831      1    2      NaN      NaN    NaN    NaN     NaN 1831-02-01   
2  1831      1    3      NaN      NaN    NaN    NaN     NaN 1831-03-01   

     dmy float  
0  1831.002732  
1  1831.087432  
2  1831.163934

#### Temperature 1941-2022

In [384]:
df_irtemprain_1941_2022 = pd.read_csv('data/ire/irraintemp1941-2023dubairp.csv', skiprows = 19)
df_irtemprain_1941_2022.head(3)

# get year, month, temp c

year  month  meant  maxtp  mintp  mnmax  mnmin  rain  gmin  wdsp maxgt  \
0  1941     11    6.9   14.0   -3.1    9.9    3.9  67.2  -5.7  12.0         
1  1941     12    6.5   12.7   -3.6    9.1    3.9  41.7  -7.6  12.5         
2  1942      1    4.3   11.9   -3.1    6.9    1.7  91.9  -9.5  13.1         

    sun  
0  56.1  
1  46.1  
2  72.8

In [385]:
df_irtemprain_1941_2022.rename(columns={'meant': 'temp c',
                            'rain': 'rain mm'},
                            inplace=True)
df_irtemprain_1941_2022.head(3)

year  month  temp c  maxtp  mintp  mnmax  mnmin  rain mm  gmin  wdsp maxgt  \
0  1941     11     6.9   14.0   -3.1    9.9    3.9     67.2  -5.7  12.0         
1  1941     12     6.5   12.7   -3.6    9.1    3.9     41.7  -7.6  12.5         
2  1942      1     4.3   11.9   -3.1    6.9    1.7     91.9  -9.5  13.1         

    sun  
0  56.1  
1  46.1  
2  72.8

In [386]:
df_irtemprain_1941_2022['dmy'] = (df_irtemprain_1941_2022['month'].astype(str) + '-' + 
                               df_irtemprain_1941_2022['year'].astype(str))
df_irtemprain_1941_2022.head(3)

year  month  temp c  maxtp  mintp  mnmax  mnmin  rain mm  gmin  wdsp maxgt  \
0  1941     11     6.9   14.0   -3.1    9.9    3.9     67.2  -5.7  12.0         
1  1941     12     6.5   12.7   -3.6    9.1    3.9     41.7  -7.6  12.5         
2  1942      1     4.3   11.9   -3.1    6.9    1.7     91.9  -9.5  13.1         

    sun      dmy  
0  56.1  11-1941  
1  46.1  12-1941  
2  72.8   1-1942

In [387]:
df_irtemprain_1941_2022['dmy'] = pd.to_datetime(df_irtemprain_1941_2022['dmy'], format='mixed')

df_irtemprain_1941_2022['dmy float'] =  np.where(df_irtemprain_1941_2022['dmy'].dt.is_leap_year == True,
                                              
                                        # normal year, 365 days
                                        df_irtemprain_1941_2022['dmy'].dt.year + df_irtemprain_1941_2022['dmy'].dt.day_of_year / 365, 

                                        # 366 for a leap year
                                        df_irtemprain_1941_2022['dmy'].dt.year + df_irtemprain_1941_2022['dmy'].dt.day_of_year / 366 
                                        )
df_irtemprain_1941_2022.head(3) 

year  month  temp c  maxtp  mintp  mnmax  mnmin  rain mm  gmin  wdsp maxgt  \
0  1941     11     6.9   14.0   -3.1    9.9    3.9     67.2  -5.7  12.0         
1  1941     12     6.5   12.7   -3.6    9.1    3.9     41.7  -7.6  12.5         
2  1942      1     4.3   11.9   -3.1    6.9    1.7     91.9  -9.5  13.1         

    sun        dmy    dmy float  
0  56.1 1941-11-01  1941.833333  
1  46.1 1941-12-01  1941.915301  
2  72.8 1942-01-01  1942.002732

#### 2.3.2 Ireland Precipitation

#### Precipitation 1711–2016

In [388]:
df_irrain_1711_2016 = pd.read_csv('data/ire/irrainIOI_1711-2016.csv')
df_irrain_1711_2016.head(3)

Year  Month  Median montly series
0  1711      1                  16.4
1  1711      2                  73.1
2  1711      3                 121.2

In [389]:
df_irrain_1711_2016.rename(columns={'Year': 'year',
                            'Month': 'month',
                            'Median montly series': 'rain mm'},
                            inplace=True)
df_irrain_1711_2016.head(3)

year  month  rain mm
0  1711      1     16.4
1  1711      2     73.1
2  1711      3    121.2

In [390]:
df_irrain_1711_2016['dmy'] = (df_irrain_1711_2016['month'].astype(str) + '-' + 
                               df_irrain_1711_2016['year'].astype(str))
df_irrain_1711_2016.head(3)

year  month  rain mm     dmy
0  1711      1     16.4  1-1711
1  1711      2     73.1  2-1711
2  1711      3    121.2  3-1711

In [391]:
df_irrain_1711_2016['dmy'] = pd.to_datetime(df_irrain_1711_2016['dmy'], format='mixed')

df_irrain_1711_2016['dmy float'] =  np.where(df_irrain_1711_2016['dmy'].dt.is_leap_year == True,
                                              
                                        # normal year, 365 days
                                        df_irrain_1711_2016['dmy'].dt.year + df_irrain_1711_2016['dmy'].dt.day_of_year / 365, 

                                        # 366 for a leap year
                                        df_irrain_1711_2016['dmy'].dt.year + df_irrain_1711_2016['dmy'].dt.day_of_year / 366 
                                        )
df_irrain_1711_2016.head(3) 

year  month  rain mm        dmy    dmy float
0  1711      1     16.4 1711-01-01  1711.002732
1  1711      2     73.1 1711-02-01  1711.087432
2  1711      3    121.2 1711-03-01  1711.163934

#### Precipitation 2017-2022

Data for the period above is taken from df_irtemprain_1941_2022.csv.

### 2.4 Data Fusion To csv And json

### 2.5 Analysis Of CO<sub>2</sub>, Temperature, CH<sub>4</sub> and Precipitation

frequency analysis of a time-based signal
fft function Fast Fourier Transformation FFT-to analyse time series - extracting frequency info abt a time signal

### 2.6 Global Temperature Anomaly Prediction

### 2.7 Accelerated Warming

## 4.0 Conclusions

## 5.0 References

In [392]:
***

## End of Project 2

SyntaxError: invalid syntax (690207592.py, line 1)

https://www.climate.gov/news-features/understanding-climate/climate-change-atmospheric-carbon-dioxide

https://theconversation.com/the-three-minute-story-of-800-000-years-of-climate-change-with-a-sting-in-the-tail-73368

https://www.researchgate.net/publication/340121184_Forty_years_of_ice-core_records_of_CO2

https://www.beyondepica.eu/en/

https://doi.pangaea.de/10.1594/PANGAEA.710901

better Co2 data than this: https://www.ncei.noaa.gov/pub/data/paleo/icecore/antarctica/antarctica2015co2composite.txt

better temp data than this: https://www.pangaea.de/?q=temperature+anomaly+ice+core

Carbon dioxide pumphandle - 2022: https://www.youtube.com/watch?v=gbxEsG8g6BA

Co2, CH4, N2o, SF6: https://gml.noaa.gov/ccgg/trends/data.html

methane 800kyr: https://www.ncei.noaa.gov/access/paleo-search/?dataTypeId=7

1. merge ice core x Mauna Loa data for temp and co2 from 800kyrs to now- some will hv time overlap, look at how the composite record from both these experiments was formed, overlay all the graphs, some will have time overlap. overlay them all on the same axis, with the composite, how did they decide with overlapping years, which one to take? and how much agreement between all diff datasets, what temporal overlap in time between them all? - RECREATE THE GRAPH FROM coluthi.xlx 80kyr of CO2 (data is from coluthi Vostok-TD-DomeC tab). temp is from 

beware of temp unit. 41586_2008_BFnature06949_MOESM31_ESM, Jouzel's data (epicaDC.deutemp.EDC3-AICC), 191.8 VS186ya, temp in kelvin - SO how did they get the temp anomaly? was all these expressed in Celcius, and these are averages calculated over the entire period or what's the offset to change 218.3 (absolute temp in kelvin in deuttemp) to 0.88, 1.84 (temp anomaly in BFnature) for year 38,46 in BFnature datafile? how was that transformation done?

other feature: ncei.noaa.gov/products/paleoclimatology, ncei.noaa.gov/access/paleo-search, CH4 https://euanmearns.com/the-vostok-ice-core-temperature-co2-and-ch4  or sea ice https://nsidc.org/data/seaice_index, https://www.climate.gov/news-features/event-tracker/2023-antarctic-sea-ice-winter-maximum-lowest-record-wide-margin 2023 record low, way outside std, how many std from mean is it out, estimate how much of a fluke it is VS https://earth.gsfc.nasa.gov/cryo/data/arcticantarctic-sea-ice-time-series or total solar irradiance (look in paleo)

in Irish context: https://www.rte.ie/brainstorm/2023/1020/1412054-human-induced-climate-change-irish-weather/, https://www.maynoothuniversity.ie/icarus/news/icarus-research-identifies-climate-change-weather-records-0 temp and precipitation data from long-term trends in daily extreme air temperature indices in ireland from 1885 to 2018 https://www.sciencedirect.com/science/article/pii/S2212094722000469- https:www.met.ie/climate/available-data/long-term-data-sets/ reproduce some of the analysis in Murphy paper

relationship between all variables-correlation, does one lead/lag to the other, 

Just for temp: linear regression - predict global temp anomaly over next few decases, synthesise data and compare to published climate models if atmospheric CO2 trends continue

Analysis:

traditionalist vs accelerationists- https://twitter.com/ryankatzrosene/status/1710108929789759812 look back at source data and see with CO2, esp temp, are we seeing linear/nonlinear increase ? Mann: warming is linear, but effects from warming is not. https://edition.cnn.com/2023/11/20/climate/2-degree-warming-limit-record-copernicus-climate-int/index.html, 

comment: are the data that u see outside of normal noise or within? if outside, there's a signal happening in data, not just some random effects of noise

***

CO2 data from Luthi et al. 2008: https://static-content.springer.com/esm/art%3A10.1038%2Fnature06949/MediaObjects/41586_2008_BFnature06949_MOESM31_ESM.xls

CO2 data from IPCC: paper: https://www.ipcc.ch/report/ar6/wg1/downloads/report/IPCC_AR6_WGI_Chapter01_SM.pdf
                data: https://agupubs.onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1002%2F2014GL061957&file=grl52461-sup-0003-supplementary.xls

Temp Jouzel - epicaDC.deuttemp.EDC3-AICC: https://vlegalwaymayo.atu.ie/pluginfile.php/1171393/mod_resource/content/0/epicaDC.deuttemp.EDC3-AICC.xls

Temperature Data from NOAA: https://www.ncei.noaa.gov/access/paleo-search/study/6080

Irish Datasets: https://www.met.ie/climate/available-data/long-term-data-sets/